Testing.Rmd

```diff
@@ -452,27 +1097,109 @@ In short, there are times when, instead of getting a failure, you just want to s
+```
+
+We have mixed feelings about this practice.
+Within the realm of top-level code in test files, this is probably on the "more acceptable" / "least offensive" end.
+But once a test file does not fit entirely on your screen, it creates an implicit yet easy-to-miss connection between the skip(s) and individual tests.
+
+At the complete other extreme, another practice that can make sense in very specific situations is to build a `skip()` directly into a package function.
+Here's an example from pkgdown, in `pkgdown::convert_markdown_to_html()`, which absolutely, positively cannot work if the Pandoc tool is unavailable:
+
+<https://github.com/r-lib/pkgdown/blob/98d5a5c735eb244cb98b2e6bab1d54bb27c0af95/R/markdown.R#L95-L106>
+
+```{r eval = FALSE}
+# in r-lib/pkgdown/R/markdown.R
+convert_markdown_to_html <- function(in_path, out_path, ...) {
+  if (rmarkdown::pandoc_available("2.0")) {
+    from <- "markdown+gfm_auto_identifiers-citations+emoji+autolink_bare_uris"
+  } else if (rmarkdown::pandoc_available("1.12.3")) {
+    from <- "markdown_github-hard_line_breaks+tex_math_dollars+tex_math_single_backslash+header_attributes"
+  } else {
+    if (is_testing()) {
+      testthat::skip("Pandoc not available")
+    } else {
+      abort("Pandoc not available")
+    }
+  }
+  
+  rmarkdown::pandoc_convert(...)
+  invisible()
+}
+```
+
+If Pandoc is not available when `convert_markdown_to_html()` executes, it throws an error *unless* it appears to part of a test run, in which case the test is skipped.
+This is an alternative to implementing a custom skipper, e.g. `skip_if_no_pandoc()`, and inserting it into many of pkgdown's tests.
+
+`testthat::is_testing()` is a small convenience function that checks the value of the `TESTTHAT` environment variable, which will be true when testthat is running tests:
+
+```{r eval = FALSE}
+is_testing <- function() {
+  identical(Sys.getenv("TESTTHAT"), "true")
+}
+```
+
+You usually want to avoid a runtime dependency on testthat, i.e. you won't want to use testthat in the code below `R/`.
+In that case, you should inline the source of `is_testing()` into your package, in whatever location you are using for any other test helper functions.
+
+<!-- Of course, the example above actually DOES have a runtime dependency on testthat, because of the call to `testthat::skip()`.
+
+So pkgdown might as well call `testthat::is_testing(), instead of inlining it.
+
+pkgdown lists testthat in Suggests, not Imports, so that is also interesting to me. Hmmm.
+-->
+
+Advantages of the embedded `skip()` approach: Feels elegant and concise.
+
+Disadvantages of this approach: Seems to violate "Good production code is well-factored; good test code is obvious", i.e. when you're focused on the tests, the existence of this skip is easy to miss.
+
+<!-- I want to talk about the danger of skipping, i.e. when you think everything's "green", but actually you're not running any / many tests, because some pre-condition is not met. -->
```

Comment:
One challenge with skips is that they are currently completely invisible — if you automatically skip too many tests it's easy to fool yourself that all your tests are passing when in fact they're just being skipped. In an ideal world, your CI/CD would make it easy to see how many tests are being skipped and how that changes over time. 

Good practice to actually dig into the R CMD check results and look at the test results and make sure the skips are as you expect. But this tends to be something you have to learn through experience.